test/test_quantized.py

```diff
@@ -121,5 +171,81 @@ def test_qsumrelu_different_qparams(self):
         np.testing.assert_equal(qC, qC_hat.int_repr())
 
 
-if __name__ == '__main__':
+@unittest.skipIf(
+    TEST_WITH_UBSAN or not torch.fbgemm_is_cpu_supported(),
+    " Quantized FC requires FBGEMM. FBGEMM does not play"
+    " well with UBSAN at the moment, so we skip the test if"
+    " we are in a UBSAN environment.",
+)
+class TestQuantizedFC(unittest.TestCase):
+    """Tests the correctness of the quantized::fc op."""
+
+    def test_qfc(self):
+        qfc_prepack = torch.ops.quantized.fbgemm_linear_prepack
+        qfc = torch.ops.quantized.fbgemm_linear
+
+        batch_size = 4
+        input_channels = 16
+        output_channels = 8
+
+        X_scale = 1.5
+        X_zp = 5
+        X_value_min = 0
+        X_value_max = 225
+        X_q = np.round(
+            np.random.rand(batch_size, input_channels) * (X_value_max - X_value_min)
+            + X_value_min
+        ).astype(np.uint8)
+
+        W_scale = 0.4
+        W_zp = 2
+        W_value_min = -128
+        W_value_max = 127
+        W_q = np.round(
+            np.random.rand(output_channels, input_channels)
+            * (W_value_max - W_value_min)
+            + W_value_min
+        ).astype(np.int8)
+
+        b_q = np.round(np.random.randn(output_channels) * 10 - 10).astype(np.int32)
+
+        # Make sure X_scale * W_scale * input_channels * X_value_max * W_value_max
```

Comment:
Why is this condition needed? Shouldn't the test pass for any value of Y_scale?

Reply:
This is not a necessary condition. As you mentioned, the test should pass for any values of Y_scale. However, when Y_scale is too small, we will get a result matrix Y_q/Y_q_ref with almost all entries 255. Then the assert_equal comparisons in the unit test might not make too much sense.

Reply:
Here are the results for different settings of Y_scale. Finally we choose Y_scale = 125.1234 for better test coverage. We should use python hypothesis for better testing in the future.

1. Y_scale = 125.1234

Y_q:
[[100  87   0 255  24   0 115 158]
 [108 107   0 195   0   0  38 255]
 [119  38   0 100  16 144  48 201]
 [ 94 121   0 255   0   0  84 255]]
Y_q_ref:
[[100  87   0 255  24   0 115 158]
 [108 107   0 195   0   0  38 255]
 [119  38   0 100  16 144  48 201]
 [ 94 121   0 255   0   0  84 255]]
Y_q_ref2:
[[100  87   0 255  24   0 115 158]
 [108 107   0 195   0   0  38 255]
 [119  38   0 100  16 144  48 201]
 [ 94 121   0 255   0   0  84 255]]


2. Y_scale = 1345.1234

Y_q:
[[ 0 16  0 12  1  1  0 21]
 [ 0 13  0  0  0  4  0 31]
 [ 0 17  0  6 10 15  0  0]
 [ 0  9  0 28  7  5  0 12]]
Y_q_ref:
[[ 0 16  0 12  1  1  0 21]
 [ 0 13  0  0  0  4  0 31]
 [ 0 17  0  6 10 15  0  0]
 [ 0  9  0 28  7  5  0 12]]
Y_q_ref2:
[[ 0 16  0 12  1  1  0 21]
 [ 0 13  0  0  0  4  0 31]
 [ 0 17  0  6 10 15  0  0]
 [ 0  9  0 28  7  5  0 12]]


3. Y_scale = 13.1234

Y_q:
[[255 255 252 255   0   0   0   0]
 [118 255   0 229   0   0  36   0]
 [255 255 255 111   0   0   0   0]
 [255 255   0 135   0   0   0   0]]
Y_q_ref:
[[255 255 252 255   0   0   0   0]
 [118 255   0 229   0   0  36   0]
 [255 255 255 111   0   0   0   0]
 [255 255   0 135   0   0   0   0]]
Y_q_ref2:
[[255 255 252 255   0   0   0   0]
 [118 255   0 229   0   0  36   0]
 [255 255 255 111   0   0   0   0]
 [255 255   0 135   0   0   0   0]]